### PR TITLE
[Core] Add options(...) function to @ray.remote stubs

### LINKED ISSUE #47329: [Core] mypy failure when `.options(...)` is used on functions decorated with `@ray.remote`
### What happened + What you expected to happen

`RemoteFunction[N]` stubs don't include a `.options` function, leading to spurious type failures.

### Versions / Dependencies

Nightly Ray, Python 3.10, Ubuntu 22.04

### Reproduction script

```
import ray

@ray.remote
def fn(x: int) -> int:
    return 2 * x

def main() -> None:
    fn_with_cpus = fn.options(num_cpus=1)
    print(ray.get(fn_with_cpus.remote(4)))

if __name__ == "__main__":
    ray.init()
    main()
    ray.shutdown()
```

produces: `error: "RemoteFunction0[int, int]" has no attribute "options"  [attr-defined]`

### Issue Severity

Low: It annoys or frustrates me.

### PR DESCRIPTION
**Pull request by milesvant**
## Why are these changes needed?

Removes incorrect `mypy` failures resulting from using `@ray.remote`.

On my reproduction script:
```
import ray


@ray.remote
def fn(x: int) -> int:
    return 2 * x

def main() -> None:
    fn_with_cpus = fn.options(num_cpus=1)
    print(ray.get(fn_with_cpus.remote(4)))

if __name__ == "__main__":
    ray.init()
    main()
    ray.shutdown()
```
with nightly Ray, `mypy script.py` produces `error: "RemoteFunction0[int, int]" has no attribute "options"  [attr-defined]` whereas my branch produces no error.

## Related issue number

Closes https://github.com/ray-project/ray/issues/47329.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [x] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

### CODE REVIEW

**Comment by milesvant** [comment]
Seems like the failing `rllib` tests are unrelated to me

**Comment by anyscalesam** [comment]
@ruisearch42 since this is on the worker side of the house can you PTAL? @can-anyscale any recommendation on how we can skip/pass the seemingly unrelated RLLib failing tests?

TY for the contribution @milesvant

File: python/ray/_private/worker.py
Comment (ruisearch42) [inline python/ray/_private/worker.py:156]: Question: how do we keep these in sync with the actual args? In other words, can/should we just use `def options(self, **task_options)` ?

File: python/ray/_private/worker.py
Comment (ruisearch42) [inline python/ray/_private/worker.py:136]: I understand that generally it is better to have proper type annotations. Just wondering in this particular case, are you trying to achieve something specific? What is the use case that you were trying to enable? IIUC currently Ray runs mypy on a limited number of files.

**Review by ruisearch42 — COMMENTED** [review]
Thanks for the contribution.

Left a couple questions, but otherwise LGTM.

File: python/ray/_private/worker.py
Comment (milesvant) [reply python/ray/_private/worker.py]: That is an improvement, thank you

Comment (milesvant) [reply python/ray/_private/worker.py]: Confirms that my reproduction script above still passes typecheck with this change

**Review by ruisearch42 — APPROVED** [review]

**Comment by milesvant** [comment]
@ruisearch42 What can I do to get this merged? I think the failing tests are flaky

**Comment by ruisearch42** [comment]
@rkooo567 Does this look good to you? If so, we can ask CI team to force merge.

**Comment by milesvant** [comment]
@ruisearch42 @rkooo567 bump

**Comment by ruisearch42** [comment]
@milesvant can you merge master and retry?

**Comment by milesvant** [comment]
@ruisearch42 `microcheck` now green

**Comment by ruisearch42** [comment]
@jjyao @rynewang (committers) Can you help merge this unless you have more comments?

**Comment by milesvant** [comment]
@ruisearch42 @jjyao @rynewang bump

**Pull request merged**
